cachi2/core/package_managers/npm.py

```diff
@@ -192,6 +387,13 @@ def fetch_npm_source(request: Request) -> RequestOutput:
         for dependency in info["dependencies"]:
             components.append(Component.from_package_dict(dependency))
 
+    npm_deps_dir = request.output_dir.join_within_root("deps", "npm")
+    npm_deps_dir.path.mkdir(parents=True, exist_ok=True)
+
+    # Download dependencies via resolved URLs and return download_paths for updating
+    # package-lock.json with local file paths
+    _get_npm_dependencies(npm_deps_dir, info["dependencies_to_download"])
```

Comment:
So I opened this in vim and my pyright LSP found a bug I missed before

![image](https://github.com/containerbuildsystem/cachi2/assets/38193343/34645170-7e1b-4dfd-8c30-a30703f2e52f)

But here's the bigger issue:
```python
for package in request.npm_packages:
    info = _resolve_npm(...)

_get_npm_dependencies(npm_deps_dir, info["dependencies_to_download"])
```
^ if the user specifies more than one npm package, we only download dependencies for the last one

Reply:
OMG, I need to switch to vim ... :( I made a fix and updated unit test for handling multiple npm packages https://github.com/containerbuildsystem/cachi2/pull/152/commits/87fb44821883bcdfcf7f59a5120069cf8716d34a but I have no idea why mypy is failing and what does he want from me :( 
```
tests/unit/package_managers/test_npm.py:391: error: Invalid index type "str" for "str"; expected type "Union[SupportsIndex, slice]"  [index]
```

Reply:
ou I'll check after lunch if we should move npm/deps folder creation outside of for `npm_deps_dir.path.mkdir(parents=True, exist_ok=True)` I believe yes

Reply:
you have `resolved_packages` annotated as
```python
    resolved_packages: dict[str, Any],
```
but they are
```python
            [
                {
                    "package": {"name": "foo", "version": "1.0.0"},
                    "dependencies": [{"name": "bar", "version": "2.0.0"}],
                    "dependencies_to_download": {
                        "https://some.registry.org/bar/-/bar-2.0.0.tgz": {
                            "integrity": "sha512-JCB8C6SnDoQf",
                            "name": "bar",
                            "version": "2.0.0",
                        }
                    },
                    "package_lock_file": ProjectFile(abspath="/some/path", template="some text"),
                },
            ],

```

Reply:
mypy is complaining because:

```python
for r in resolved_packages:  # resolved_packages: dict[str, Any] => r: str
    r["dependencies_to_download"]  # "some_string"["dependencies_to_download"]
```

Reply:
> ou I'll check after lunch if we should move npm/deps folder creation outside of for `npm_deps_dir.path.mkdir(parents=True, exist_ok=True)` I believe yes

doesn't matter too much :shrug:

Reply:
Nice Catch!! Thank you Adam.

Reply:
TIL about Pyright. There's a VSCode extension as well, Lada.

Reply:
Ah, but it's not available for the open source version https://github.com/microsoft/pylance-release/blob/main/FAQ.md#can-i-load-pylance-in-code--oss

Reply:
Pylance is what I currently use, but it did not point this specific error out.